app/constants/mods.py

```diff
@@ -48,228 +39,194 @@ class Mods(IntFlag):
+            mods &= ~cls.HALFTIME  # (DT|NC)HT
 
-        if self & (Mods.NOFAIL | Mods.RELAX | Mods.AUTOPILOT):
-            if self & Mods.SUDDENDEATH:
-                self &= ~Mods.SUDDENDEATH  # (NF|RX|AP)SD
-            if self & Mods.PERFECT:
-                self &= ~Mods.PERFECT  # (NF|RX|AP)PF
+        if (mods & cls.EASY) and (mods & cls.HARDROCK):
+            mods &= ~cls.HARDROCK  # (EZ)HR
 
-        if self & (Mods.RELAX | Mods.AUTOPILOT):
-            if self & Mods.NOFAIL:
-                self &= ~Mods.NOFAIL  # (RX|AP)NF
+        if (mods & (cls.NOFAIL | cls.RELAX | cls.AUTOPILOT)) and (
+            mods & cls.SUDDENDEATH
+        ):
+            mods &= ~cls.SUDDENDEATH  # (NF|RX|AP)SD
+        if (mods & (cls.NOFAIL | cls.RELAX | cls.AUTOPILOT)) and (mods & cls.PERFECT):
+            mods &= ~cls.PERFECT  # (NF|RX|AP)PF
 
-        if self & Mods.PERFECT and self & Mods.SUDDENDEATH:
-            self &= ~Mods.SUDDENDEATH  # PFSD
+        if (mods & (cls.RELAX | cls.AUTOPILOT)) and (mods & cls.NOFAIL):
+            mods &= ~cls.NOFAIL  # (RX|AP)NF
 
-        # 2. remove mode-unique mods from incorrect gamemodes
-        if mode_vn != 0:  # osu! specific
-            self &= ~OSU_SPECIFIC_MODS
+        if (mods & cls.PERFECT) and (mods & cls.SUDDENDEATH):
+            mods &= ~cls.SUDDENDEATH  # (PF)SD
 
-        # ctb & taiko have no unique mods
+        if mode_vn != 0:
+            mods &= ~OSU_SPECIFIC_MODS
 
-        if mode_vn != 3:  # mania specific
-            self &= ~MANIA_SPECIFIC_MODS
+        if mode_vn != 3:
+            mods &= ~MANIA_SPECIFIC_MODS
 
-        # 3. mode-specific mod conflictions
         if mode_vn == 0:
-            if self & Mods.AUTOPILOT:
-                if self & (Mods.SPUNOUT | Mods.RELAX):
-                    self &= ~Mods.AUTOPILOT  # (SO|RX)AP
+            if (mods & cls.AUTOPILOT) and (mods & (cls.SPUNOUT | cls.RELAX)):
+                mods &= ~cls.AUTOPILOT  # (SO|RX)AP
 
         if mode_vn == 3:
-            self &= ~Mods.RELAX  # rx is std/taiko/ctb common
-            if self & Mods.HIDDEN and self & Mods.FADEIN:
-                self &= ~Mods.FADEIN  # HDFI
-
-        # 4 remove multiple keymods
-        # TODO: do this better
-        keymods_used = self & KEY_MODS
+            mods &= ~cls.RELAX
+            if (mods & cls.HIDDEN) and (mods & cls.FADEIN):
+                mods &= ~cls.FADEIN  # (HD)FI
 
-        if bin(keymods_used).count("1") > 1:
-            # keep only the first
-            first_keymod = None
-            for mod in KEY_MODS:
-                if keymods_used & mod:
-                    first_keymod = mod
-                    break
+        keymods_used = mods & KEY_MODS
 
-            # remove all but the first keymod.
-            self &= ~(keymods_used & ~first_keymod)
+        if keymods_used & (keymods_used - 1):
+            # remove all but the lowest keymod
+            mods &= keymods_used & -keymods_used
```

Comment:
hey, how does this work? still not sure

Reply:
my theory (which works), as that `keymods_used & (keymods_used - 1)` checks if there are multiple keymods enabled at the same time by first ANDing `keymods_used` with `keymods_used - 1`
if there are multiple keymods enabled; `keymods_used` will have multiple bits set to 1, so `keymods_used - 1` will have at least one bit set to 1 that is not set in `keymods_used`. when these values are ANDed together, the result will be a non-zero value, indicating that multiple keymods are enabled. 

if only one keymod is enabled, `keymods_used - 1` will have all bits set to 0, so the result of `keymods_used & (keymods_used - 1)` will be 0, indicating that only one keymod is enabled, otherwise; it will execute what's in the if statement.

this appears to work flawlessly on my machine but still needs more testing from other parties to be sure